src/test/java/bio/terra/service/dataset/DatasetServiceTest.java

```diff
@@ -632,4 +634,38 @@ public void ingestPayloadDataAzure() throws Exception {
         .newJob(any(), eq(DatasetIngestFlight.class), requestCaptor.capture(), any());
     assertThat("payload is stripped out", requestCaptor.getValue().getRecords(), empty());
   }
+
+  @Test
+  public void getRetrieveDatasetRequiredActionsWithDefaults() {
+    List<IamAction> actions =
+        datasetService.getRetrieveDatasetRequiredActions(
+            List.of(
+                DatasetRequestAccessIncludeModel.SCHEMA,
+                DatasetRequestAccessIncludeModel.PROFILE,
+                DatasetRequestAccessIncludeModel.DATA_PROJECT,
+                DatasetRequestAccessIncludeModel.STORAGE));
+    assertThat("There is only one action", actions.size(), equalTo(1));
+    assertThat(
+        "The only required action is reader", actions.contains(IamAction.READ_DATASET), is(true));
+  }
+
+  @Test
+  public void getRetrieveDatasetRequiredActionsWithSnapshotBuilderConfig() {
+    List<IamAction> actions =
+        datasetService.getRetrieveDatasetRequiredActions(
+            List.of(
+                DatasetRequestAccessIncludeModel.SCHEMA,
+                DatasetRequestAccessIncludeModel.PROFILE,
+                DatasetRequestAccessIncludeModel.DATA_PROJECT,
+                DatasetRequestAccessIncludeModel.SNAPSHOT_BUILDER_CONFIG));
+    assertThat("There is only one action", actions.size(), is(2));
+    assertThat(
+        "The only required action is reader",
+        actions.contains(IamAction.READ_DATASET),
+        equalTo(true));
+    assertThat(
+        "The only required action is reader",
+        actions.contains(IamAction.VIEW_SNAPSHOT_BUILDER_SETTINGS),
+        equalTo(true));
```

Comment:
Thanks for adding some tests here.

These three assertions can be consolidated as follows, and the same approach is applicable to the test above:

```
assertThat(actions, containsInAnyOrder(IamAction.READ_DATASET, IamAction.VIEW_SNAPSHOT_BUILDER_SETTINGS));
```

I like [Matchers.contains](https://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#contains(E...)) to verify that a collection contains exactly what I think it will in the specified order.  [Matchers.containsInAnyOrder](https://hamcrest.org/JavaHamcrest/javadoc/1.3/org/hamcrest/Matchers.html#containsInAnyOrder(T...)) is an order-agnostic matcher that otherwise behaves the same.  With both, you get size verification for free.

When writing tests, I like to consider what the developer debugging a failure would get as far as information to guide their investigation.  With `contains` and `containsInAnyOrder`, they would get the full actual collection and be able to figure out what's missing.

Versus failures on the existing assertions:
- When matching the size, a failure indicates that there are more / fewer actions than expected, but they won't be shown what they are
- The other two assertions failing would present the user with "expected true but got false".  I find these messages tough to debug, forcing you to dig and rely on the assertion's message.

IMO, the consolidated action is clear enough as to not warrant an additional message, but that's a matter of opinion.

Reply:
Thank you for the info! Yeah, `Matchers.contains` definitely seems better.